contrib/gogearbox/gearbox.v1/gearbox_test.go

```diff
@@ -0,0 +1,271 @@
+
+		Middleware(WithServiceName("gb"), WithStatusCheck(customErrChecker))(gb)
+		spans := mt.FinishedSpans()
+		assert.Len(spans, 1)
+		span := spans[0]
+		assert.Equal(strconv.Itoa(code), span.Tag(ext.HTTPCode))
+		assert.Nil(span.Tag(ext.Error))
+	})
+}
+
+// Test that users can customize how resource_name is determined
+func TestCustomResourceNamer(t *testing.T) {
+	assert := assert.New(t)
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	reqctx := newReqCtx(200)
+	gb := &GearboxContextMock{requestCtx: reqctx}
+
+	customRsc := "custom resource"
+	namer := func(gctx gearbox.Context) string {
+		return customRsc
+	}
+
+	Middleware(WithResourceNamer(namer))(gb)
+	spans := mt.FinishedSpans()
+	assert.Len(spans, 1)
+	span := spans[0]
+	assert.Equal(customRsc, span.Tag(ext.ResourceName))
+}
+
+// Test that the trace middleware passes the context off to the next handler in the req chain even if the request is not instrumented
+// MTOFF: I can't create a real http request, but I can simulate middleware being run
+// by checking that the context's `index` has been incremented
+func TestWithIgnoreRequest(t *testing.T) {
+	assert := assert.New(t)
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	reqctx := newReqCtx(200)
+	// explicitly giving gb.index a value, so we can see it increment clearly
+	gb := &GearboxContextMock{requestCtx: reqctx, index: 0}
+
+	ignoreResources := func(c gearbox.Context) bool {
+		return strings.HasPrefix(string(c.Context().URI().Path()), "/any")
+	}
+	Middleware(WithIgnoreRequest(ignoreResources))(gb)
+	assert.Len(mt.FinishedSpans(), 0)
+	assert.Equal(1, gb.index)
+}
+func TestPropagation(t *testing.T) {
+	t.Run("inject-extract", func(t *testing.T) {
+		assert := assert.New(t)
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		reqctx := newReqCtx(200)
+		gb := &GearboxContextMock{requestCtx: reqctx}
+		fcc := &FasthttpContextCarrier{gb.Context()}
+
+		pspan := tracer.StartSpan("test")
+		err := tracer.Inject(pspan.Context(), fcc)
+		if err != nil {
+			t.Fatalf("Trace injection failed")
+		}
+		Middleware(WithServiceName("gb"))(gb)
+		sctx, err := tracer.Extract(fcc)
+		if err != nil {
+			t.Fatalf("Trace extraction failed")
+		}
+		assert.Equal(sctx.TraceID(), pspan.Context().TraceID())
+		assert.Equal(sctx.SpanID(), pspan.Context().SpanID())
+	})
+	t.Run("req-context", func(t *testing.T) {
+		assert := assert.New(t)
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		reqctx := newReqCtx(200)
+		gb := &GearboxContextMock{requestCtx: reqctx}
+		fcc := &FasthttpContextCarrier{gb.Context()}
+
+		pspan := tracer.StartSpan("test")
+		err := tracer.Inject(pspan.Context(), fcc)
+		if err != nil {
+			t.Fatalf("Trace injection failed")
+		}
+		Middleware(WithServiceName("gb"))(gb)
+		span, ok := tracer.SpanFromContext(gb.Context())
+		assert.True(ok)
+		assert.Equal(span.(mocktracer.Span).TraceID(), pspan.(mocktracer.Span).TraceID())
+		assert.Equal(span.(mocktracer.Span).ParentID(), pspan.(mocktracer.Span).SpanID())
+	})
+}
+
```

Comment:
Gearbox doesn't have any testing utils nor any implementation of some kind of in-memory listener. So I couldn't spin up a client & server. Therefore I chose to mock Gearbox.Context.